File: src/main/java/no/novari/flyt/mapping/model/configuration/ValueMapping.java
package no.novari.flyt.mapping.model.configuration;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class ValueMapping {

    public enum Type {
        STRING, URL, BOOLEAN, DYNAMIC_STRING, FILE, VALUE_CONVERTING
    }

    private final Type type;

    private final String mappingString;

}

File: src/main/kotlin/no/novari/flyt/mapping/model/configuration/ValueMapping.kt
package no.novari.flyt.mapping.model.configuration

data class ValueMapping(
    val type: Type,
    val mappingString: String? = null,
) {
    enum class Type {
        STRING,
        URL,
        BOOLEAN,
        DYNAMIC_STRING,
        FILE,
        VALUE_CONVERTING,
    }

    class Builder {
        private var type: Type? = null
        private var mappingString: String? = null

        fun type(type: Type) = apply { this.type = type }

        fun mappingString(mappingString: String?) = apply { this.mappingString = mappingString }

        fun build(): ValueMapping {
            return ValueMapping(
                type = requireNotNull(type) { "type is required" },
                mappingString = mappingString,
            )
        }
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }
}
